### PR TITLE
fv3fit: clean up sample dim arg

### PR DESCRIPTION
**Pull request by AnnaKwa**
This removes the `sample_dim_name` arg from the initialization of most of the models where it was just an internal implementation detail used to stack and unstack data. The one exception to this is in the BPTT `StepwiseModel` where dataset input to `integrate_stepwise` already has a sample dimension which might not match the `fv3fit._shared.SAMPLE_DIM_NAME` used in other models.

### CODE REVIEW

File: external/fv3fit/fv3fit/keras/_models/recurrent.py [hunk 1]
```diff
@@ -546,6 +549,17 @@ class StepwiseModel(PureKerasModel):
     include air_temperature and specific_humidity
     """
 
+    def __init__(
+        self,
+        input_variables: Iterable[Hashable],
+        output_variables: Iterable[Hashable],
+        output_metadata: Iterable[Dict[str, Any]],
+        model: tf.keras.Model,
+        sample_dim_name: str,
```
Comment (AnnaKwa) [inline external/fv3fit/fv3fit/keras/_models/recurrent.py:558]: This is the one model that still has sample dim as an arg, since it looks like its method to integrate forward in time works on a dataset that can have any name for its sample dim.

File: external/fv3fit/fv3fit/keras/_models/recurrent.py [hunk 2]
```diff
@@ -430,12 +433,12 @@ def get_target_state(self, X):
     @property
     def predictor_model(self) -> "StepwiseModel":
         predictor_model = StepwiseModel(
-            self.sample_dim_name,
             tuple(self.input_packer.pack_names)
             + tuple(self.prognostic_packer.pack_names),
             self.output_variables,
             self._stepwise_output_metadata,
             self.predict_keras_model,
+            self.sample_dim_name,
```
Comment (mcgibbon) [inline external/fv3fit/fv3fit/keras/_models/recurrent.py:441]: suggestion: if you update these to use keyword arguments for the variables (just here, you don't need to change the call signature) we can bugs from mis-ordered arguments.

File: external/fv3fit/fv3fit/keras/_models/recurrent.py (hunk 1 above)
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/recurrent.py:558]: This is fine.

File: external/fv3fit/fv3fit/keras/_models/recurrent.py [hunk 3]
```diff
@@ -588,3 +602,39 @@ def integrate_stepwise(self, ds: xr.Dataset) -> xr.Dataset:
...
+                f.write(
+                    yaml.dump(
+                        {
+                            "input_variables": self.input_variables,
+                            "output_variables": self.output_variables,
+                            "output_metadata": self._output_metadata,
+                            "sample_dim_name": self.sample_dim_name,
+                        }
+                    )
+                )
```
Comment (mcgibbon) [inline external/fv3fit/fv3fit/keras/_models/recurrent.py:640]: Why is this appearing as added code? Did you add it? If so then why, it seems unrelated?

File: workflows/prognostic_c48_run/tests/_regtest_outputs/test_machine_learning.test_ml_steppers_regression_checksum[tendencies].out [hunk 4]
```diff
@@ -4,9 +4,9 @@
     - - dQ2
       - 8e405c2b4f212a44d5604ece97b291e2
     - - dQu
-      - 4d9fdc296ac232d82308a2f8e3d5e444
+      - 6da7a5b20608af0b0a30bc891861e807
     - - dQv
-      - 4d9fdc296ac232d82308a2f8e3d5e444
+      - 6da7a5b20608af0b0a30bc891861e807
```
Comment (mcgibbon) [inline workflows/prognostic_c48_run/tests/_regtest_outputs/test_machine_learning.test_ml_steppers_regression_checksum[tendencies].out:9]: Do you have an explanation for why these changed? I don't understand why.

File: workflows/prognostic_c48_run/tests/_regtest_outputs/test_regression.test_fv3run_diagnostic_outputs_schema[nudging].out [hunk 5]
```diff
@@ -1,10 +1,10 @@
 xarray.Dataset {
 dimensions:
-	time = 2 ;
 	tile = 6 ;
-	z = 63 ;
-	y = 12 ;
+	time = 2 ;
 	x = 12 ;
+	y = 12 ;
+	z = 63 ;
```
Comment (mcgibbon) [inline workflows/prognostic_c48_run/tests/_regtest_outputs/test_regression.test_fv3run_diagnostic_outputs_schema[nudging].out:7]: Inconsequential changes like this are fine.

**Review by mcgibbon — COMMENTED** [review]
BPTT training still uses the old behavior (uses the first dimension of air_temperature in the dataset as the sample dimension name) but that's fine - I don't know what would happen if you changed it. The main need has been to update the `fv3fit.train` endpoints to remove these args.

I need to understand why dump/load was added to StepwisePredictor and have some more idea of why the regression targets changed (or at least why it's fine) before I can approve.

File: external/fv3fit/fv3fit/keras/_models/recurrent.py (hunk 3 above)
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/recurrent.py:640]: OK.

File: workflows/prognostic_c48_run/tests/_regtest_outputs/test_regression.test_fv3run_diagnostic_outputs_schema[nudging].out (hunk 5 above)
Comment (AnnaKwa) [reply workflows/prognostic_c48_run/tests/_regtest_outputs/test_regression.test_fv3run_diagnostic_outputs_schema[nudging].out:7]: Thank you!! I was going nuts trying to think why the ConstantOutputPredictor output would have changed. I think this was it.

**Review by mcgibbon — APPROVED** [review]
LGTM
